momaland/envs/samegame/samegame.py

```diff
@@ -0,0 +1,531 @@
+        self.agents = self.possible_agents[:]
+        self.move_count = 0
+        self.max_move = board_height * board_width
+        self._check_parameters()
+        self._initialize_board()
+        self.legal_moves = self._legal_moves()
+        self.terminations = {agent: False for agent in self.agents}
+        self.truncations = {agent: False for agent in self.agents}
+        self.infos = {i: {} for i in self.agents}
+        self.rewards = {i: np.zeros(self.num_objectives) for i in self.agents}
+        self._cumulative_rewards = {i: np.zeros(self.num_objectives) for i in self.agents}
+        self.action_spaces = {i: spaces.Discrete(self.max_move) for i in self.agents}
+        self.observation_spaces = {
+            i: spaces.Dict(
+                {
+                    "observation": spaces.Box(
+                        low=0, high=1, shape=(board_height, board_width, self.gameinfo["ncolors"]), dtype=np.int8
+                    ),
+                    "action_mask": spaces.Box(low=0, high=1, shape=(self.max_move,), dtype=np.int8),
+                }
+            )
+            for i in self.agents
+        }
+        self.reward_spaces = dict(
+            zip(
+                self.agents,
+                [spaces.Box(low=0, high=self._score(board_height * board_width), shape=(self.num_objectives,))]
+                * len(self.agents),
+            )
+        )
+
+    @override
+    def render(self):
+        """Renders the environment.
+
+        In human mode, it can print to terminal, open
+        up a graphical window, or open up some other display that a human can see and understand.
+        """
+        if self.render_mode is None:
+            warn("You are calling render method without specifying any render mode.")
+            return
+
+        if self.render_mode == "human":
+            self.print_board()
+
+    @override
+    def observe(self, agent):
+        board = self.gameinfo["board"][self.gameinfo["curmove"]]
+        observation = np.stack(([np.equal(board, color + 1) for color in range(self.gameinfo["ncolors"])]), axis=2).astype(
+            np.int8
+        )
+        actions = self.legal_moves if agent == self.agent_selection else []
+        action_mask = np.zeros(self.max_move, "int8")
+        for i in actions:
+            action_mask[i] = 1
+        return {"observation": observation, "action_mask": action_mask}
+
+    @override
+    def observation_space(self, agent):
+        return self.observation_spaces[agent]
+
+    @override
+    def action_space(self, agent):
+        return self.action_spaces[agent]
+
+    @override
+    def reward_space(self, agent):
+        return self.reward_spaces[agent]
+
+    def _legal_moves(
+        self,
+    ):  # TODO at the moment, every clickable coordinate is a distinct action. an alternative would be having only 1 distinct legal action per clickable group (via a representative coordinate for the group, i.e. the one in the top right)
```

Comment:
I think the way it is now is better than the alternative suggested in the todo. The benefit of having one action per coordinate is that you can easily generate masks. If you have one action per clickable group, that's a new problem. Therefore, I would remove the todo and just not do it.

Reply:
It makes learning and searching easier, but programming harder :) Agreed